### PR TITLE
Add active/inactivate state to the Droplet module

### LINKED ISSUE #23: Power on and power off droplets
##### SUMMARY
Implement the basic, fundamental, and obvious feature of changing the power state of an existing droplet.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
digital_ocean_droplet

##### ADDITIONAL INFORMATION
This feature would allow the simple, obvious, and necessary ability to TURN ON or TURN OFF a droplet that already exists. This is probably the single most basic VPS function and really should already exist.
The feature would probably most easily be called by users via the STATE: parameter, with something like "active/inactive", poweredon/poweredoff, etc.

```yaml
- name: power on my drupal-test droplet
  digital_ocean_droplet:
    state: active
    id: "{{ my_droplet_id }}"
    oauth_token: "{{ my_do_api_token }}"
```

### PR DESCRIPTION
**Pull request by mamercad**
##### SUMMARY

Add active/inactivate state to the Droplet module, fixes #23.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

digital_ocean_droplet

##### ADDITIONAL INFORMATION

```bash
[WARNING]: You are running the development version of Ansible. You should only
run Ansible from "devel" if you are modifying the Ansible engine, or trying out
features under development. This is a rapidly changing source of code and can
become unstable at any point.
[WARNING]: running playbook inside collection community.digitalocean
[WARNING]: Unable to parse /Users/mmercado/.ansible/collections/ansible_collect
ions/community/digitalocean/localhost as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

PLAY [test] ********************************************************************

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] =>
  doctl.stdout_lines:
  - ID    Name    Public IPv4    Private IPv4    Public IPv6    Memory    VCPUs    Disk    Region    Image    VPC UUID    Status    Tags    Features    Volumes

TASK [Creating a Droplet in DigitalOcean (present)] ****************************
changed: [localhost]

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] =>
  doctl.stdout_lines:
  - ID           Name             Public IPv4       Private IPv4    Public IPv6    Memory    VCPUs    Disk    Region    Image                     VPC UUID                                Status    Tags    Features              Volumes
  - '211718525    hacktoberfest    REDACTED    10.108.0.2                     2048      1        50      nyc3      Ubuntu 20.04 (LTS) x64    588f833b-e482-4af4-b049-2f0c74b201cd    active            private_networking    '

TASK [Creating a Droplet in DigitalOcean (active)] *****************************
ok: [localhost]

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [Creating a Droplet in DigitalOcean (inactive)] ***************************
changed: [localhost]

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] =>
  doctl.stdout_lines:
  - ID           Name             Public IPv4       Private IPv4    Public IPv6    Memory    VCPUs    Disk    Region    Image                     VPC UUID                                Status    Tags    Features              Volumes
  - '211718525    hacktoberfest    REDACTED    10.108.0.2                     2048      1        50      nyc3      Ubuntu 20.04 (LTS) x64    588f833b-e482-4af4-b049-2f0c74b201cd    off               private_networking    '

TASK [Creating a Droplet in DigitalOcean (active)] *****************************
changed: [localhost]

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] =>
  doctl.stdout_lines:
  - ID           Name             Public IPv4       Private IPv4    Public IPv6    Memory    VCPUs    Disk    Region    Image                     VPC UUID                                Status    Tags    Features              Volumes
  - '211718525    hacktoberfest    REDACTED    10.108.0.2                     2048      1        50      nyc3      Ubuntu 20.04 (LTS) x64    588f833b-e482-4af4-b049-2f0c74b201cd    active            private_networking    '

TASK [Creating a Droplet in DigitalOcean (absent)] *****************************
changed: [localhost]

TASK [List all Droplets in DigitalOcean] ***************************************
changed: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] =>
  doctl.stdout_lines:
  - ID           Name             Public IPv4       Private IPv4    Public IPv6    Memory    VCPUs    Disk    Region    Image                     VPC UUID                                Status    Tags    Features              Volumes

PLAY RECAP *********************************************************************
localhost                  : ok=16   changed=10   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

The state is in parentheses of the `TASK` line (`present`, `active`, `inactive`, `absent`), look at the corresponding "Status" column of the `doctl compute drop droplet list` command output, in conjunction with the `ok/changed` result of the `TASK [Create a Droplet ...]`.

### CODE REVIEW

File: plugins/modules/digital_ocean_droplet.py [hunk 1]
```diff
@@ -313,20 +324,31 @@ def ensure_power_on(self, droplet_id):
...
+            time.sleep(min(2, end_time - time.time()))
+        self.module.fail_json(msg='Wait for droplet powering off timeout')
+
 
 def core(module):
     state = module.params.pop('state')
     droplet = DODroplet(module)
-    if state == 'present':
-        droplet.create()
+    if state == 'present' or state == 'active' or state == 'inactive':  # These states imply that we have, or create, the droplet
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_droplet.py:342]: Grammar nit: the verb tense in this comment is confusing to me and I'm not sure exactly what it's trying to say.

**Review by mpontillo — DISMISSED** [review]
LGTM; some minor nits below.

File: plugins/modules/digital_ocean_droplet.py [hunk 2]
```diff
@@ -18,8 +18,12 @@
   state:
     description:
      - Indicate desired state of the target.
+     - C(present) will create named droplet, be mindful of the C(unique_name) parameter
```
[the remark this answers is not in the document]
Comment (mamercad) [reply plugins/modules/digital_ocean_droplet.py:21]: These are supposed to end in periods as well, will fix this also.

File: plugins/modules/digital_ocean_droplet.py [hunk 3]
```diff
@@ -271,12 +273,25 @@ def get_droplet(self):
...
             droplet_data = self.get_addresses(json_data)
-            self.module.exit_json(changed=False, data=droplet_data)
+            # If state is active or inactive, ensure requested and desired power states match
+            droplet = json_data.get('droplet', None)
+            if droplet is not None:
+                droplet_id = droplet.get('id', None)
+                droplet_status = droplet.get('status', None)
+                if droplet_id is not None and droplet_status is not None:
+                    if state == 'active' and droplet_status != 'active':
+                        json_data = self.ensure_power_on(droplet_id)
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_droplet.py:288]: This part was confusing to me because you're grabbing a new `json_data`, but unless I'm missing something, I don't think that variable is being used? If that's the case, consider not overwriting the `json_data` variable with the data from the new API call. Do you need to check the result of the `ensure_power_on` or `ensure_power_off` call before reporting back to the user?

Comment (mamercad) [reply plugins/modules/digital_ocean_droplet.py]: This should be in better shape now, I've created new variables for each type of power operation. As far as checking the return goes, I'm not sure it's completely necessary -- the `ensure_power_on/off` functions will either work (and return the json response data) or the module will fail. There may be edge cases, but, for the most part, I think my changes follow the spirit of what's already there. Once we get full integration tests going (very soon), it'll be less of a leap of faith.

File: plugins/modules/digital_ocean_droplet.py [hunk 4]
```diff
@@ -350,20 +368,31 @@ def ensure_power_on(self, droplet_id):
...
+    def ensure_power_off(self, droplet_id):
+        response = self.rest.post('droplets/{0}/actions'.format(droplet_id), data={'type': 'power_off'})
+        end_time = time.time() + self.wait_timeout
+        while time.time() < end_time:
+            response = self.rest.get('droplets/{0}'.format(droplet_id))
+            json_data = response.json
+            if json_data['droplet']['status'] == 'off':
+                return json_data
+            time.sleep(min(2, end_time - time.time()))
+        self.module.fail_json(msg='Wait for droplet powering off timeout')
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_droplet.py:380]: Consider an error message such as:
```
"could not verify droplet power-off within %ds" % d
```

File: plugins/modules/digital_ocean_droplet.py [hunk 5]
```diff
@@ -350,20 +368,31 @@ def ensure_power_on(self, droplet_id):
...
 
+    def ensure_power_off(self, droplet_id):
+        response = self.rest.post('droplets/{0}/actions'.format(droplet_id), data={'type': 'power_off'})
+        end_time = time.time() + self.wait_timeout
+        while time.time() < end_time:
+            response = self.rest.get('droplets/{0}'.format(droplet_id))
+            json_data = response.json
+            if json_data['droplet']['status'] == 'off':
+                return json_data
+            time.sleep(min(2, end_time - time.time()))
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_droplet.py:379]: This line could lead to a `ValueError` in the event of clock skew. (admittedly somewhat unlikely, but you never know on a busy machine.) For example:
```
>>> time.sleep(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: sleep length must be non-negative
```
I think you should just remove the subtraction here to keep it simple. It doesn't have to be exact. (maybe document that it isn't exact if you're concerned.)
Also maybe increase the sleep to 10 seconds; hitting the API every two seconds for two minutes seems a little excessive.

**Review by mpontillo — CHANGES_REQUESTED** [review]
Let's at least fix the potential for a negative sleep before merging this one.

File: plugins/modules/digital_ocean_droplet.py (hunk 5 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_droplet.py:379]: These are definitely good suggestions -- it's not really apparent, but, this is a shameless copy of the function "above it" (flipping "on" to "off"). I'm going to turn this into a new issue.

File: plugins/modules/digital_ocean_droplet.py (hunk 4 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_droplet.py:380]: Created #65 for this.

File: plugins/modules/digital_ocean_droplet.py (hunk 5 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_droplet.py:379]: Fix in #65.

**Review by mpontillo — APPROVED** [review]
Sounds good. Ship it then.
